mautic/mautic
# Fixes tokenized mail transports for recipients without a name and form results

Please answer the following questions. 

| Q  | A
| --- | ---
| Bug fix? | y
| New feature? | n
| Related user documentation PR URL | n
| Related developer documentation PR URL | n
| Issues addressed (#s or URLs) | #2277 #1625
| BC breaks? | n
| Deprecations? | y

**Note that all new features should have a related user and/or developer documentation PR in their respective repositories.** 

### Required
#### Description:

This PR fixes two things. 

1. A bug was introduced with #2132 we didn't catch that caused the queuedRecipients array to be keyed numerically for those without a name. This broke transports like Mandrill, and now Sparkpost, because tokens were no longer found for those email addresses. 

2. Fixing this resolves #2277 but also adds support for contact token fields in the form results email.

#### Steps to test this PR:
1. Create a standalone form that includes an action to send form results. Add a {contactfield=firstname} token in the email body somewhere.
2. Configure the system's mailer to Mandrill or Sparkpost.
3. Submit the form in an anonymous browser session.
4. Check the email for token content. 

### As applicable
#### Steps to reproduce the bug:
1. Repeat the steps before applying the PR and the email should arrive without the contact token getting replaced and no form values populated.

#### List deprecations along with the new alternative:
1. This deprecates \Mautic\EmailBundle\Helper\MailHelper::useMailerTokenization(). Due to the way tokenized transports are written - tokenization had to be used regardless of the code implementing the mail send. A check was added in MailHelper::send() to auto-handle these transports so that the implementing code can use queue()/flushQueue() or send() without knowing what the transport supports.
